openedx_lti_tool_plugin/tasks.py

```diff
@@ -0,0 +1,83 @@
+"""Tasks for openedx_lti_tool_plugin."""
+import logging
+from datetime import datetime, timezone
+
+from celery import shared_task
+from django.contrib.auth import get_user_model
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+from openedx_lti_tool_plugin.edxapp_wrapper.grades_module import course_grade_factory
+from openedx_lti_tool_plugin.edxapp_wrapper.modulestore_module import modulestore
+from openedx_lti_tool_plugin.models import LtiGradedResource
+
+log = logging.getLogger(__name__)
+
+
+@shared_task(name='openedx_lti_tool_plugin.tasks.send_problem_score_update',)
+def send_problem_score_update(
+    problem_weighted_earned: str,
+    problem_weighted_possible: str,
+    user_id: str,
+    problem_id: str,
+):
+    """Send problem score update task.
+
+    Task to update the AGS score of a problem asynchronously.
+
+    Args:
+        problem_weighted_earned: Grade earned for the problem.
+        problem_weighted_possible: Grade possible for the problem.
+        user_id: Grading user id.
+        problem_id: Problem id.
+    """
+    for graded_resource in LtiGradedResource.objects.all_from_user_id(user_id=user_id, context_key=problem_id):
+        log.info('LTI AGS: Sending AGS update for problem %s with user %s', problem_id, user_id)
+        graded_resource.update_score(
+            problem_weighted_earned,
+            problem_weighted_possible,
+            datetime.now(tz=timezone.utc),
+        )
+
+
+@shared_task(name='openedx_lti_tool_plugin.tasks.send_vertical_score_update',)
+def send_vertical_score_update(
+    user_id: str,
+    course_id: str,
+    problem_id: str,
+):
+    """Send vertical score update task.
+
+    Task to obtain a vertical's accumulated grade and update the AGS score asynchronously. This is a task that would be
+    executed whenever a problem score is updated. We decided to do it this way because there is no way of telling if a
+    score of a unit was changed.
+
+    Args:
+        user_id: Grading user id.
+        course_id: Context course id string.
+        problem_id: Problem id.
+    """
+    user = get_user_model().objects.get(id=user_id)
+    problem_key = UsageKey.from_string(problem_id)
+    problem_descriptor = modulestore().get_item(problem_key)
+    vertical_key = problem_descriptor.parent
+
+    vertical_graded_resources = LtiGradedResource.objects.all_from_user_id(
+        user_id=user.id,
+        context_key=str(vertical_key),
+    )
+
+    if not vertical_graded_resources:
+        return
+
+    course_key = CourseKey.from_string(course_id)
+    course_descriptor = modulestore().get_course(course_key)
+    course_grade = course_grade_factory().read(user, course_descriptor)
+    earned, possible = course_grade.score_for_module(vertical_key)
```

Comment:
```suggestion
    earned, possible = course_grade.score_for_module(vertical_key)
    
    if not earned or not possible:
        return
```

@alexjmpb What do you think about not sending a score update if no earned value is received (or earned value is 0.0) and not possible value exists for the unit?

Reply:
Since we are sending the final grade and not the passed grade, can't it be the case that a student receives a final grade of 0.0? Say is a exam problem with only one attempt and the student fails that?